File: backoffice/src/shared/infrastructure/events/DomainEventSubscribersLocator.py
"""
 *
 * Libraries 
 *
"""

import inspect
import typing
from types                  import ModuleType
from src.shared.domain      import DomainEventSubscriber
from ..ContextModulesFinder import ContextModulesFinder

"""
 *
 * Class
 *
"""

class DomainEventSubscribersLocator:

    """
     *
     * Methods 
     *
    """

    @classmethod
    def locate( cls, eventClass : type ) -> list[type]:
        # Variables
        subscriberClasses  : list[type]
        applicationModules : list[ModuleType]
        satisfiedClasses   : list[type]
        # Code
        applicationModules = ContextModulesFinder.findApplicationModules()
        subscriberClasses  = []
        for applicationModule in applicationModules:
            satisfiedClasses = list( map( lambda members: members[1], inspect.getmembers( 
                applicationModule, 
                lambda anyClass: cls.__isSubscriberClass( anyClass, eventClass )                    
            ) ) )            
            subscriberClasses.extend( satisfiedClasses )
        return subscriberClasses
    
    @classmethod
    def __getEventClass( cls, eventSubscriberClass : type ) -> type:
        return typing.get_args( eventSubscriberClass.__orig_bases__[0] )[0]
    
    @classmethod
    def __isSubscriberClass( cls, anyClass : object, eventClass : type ) -> bool:
        return inspect.isclass( anyClass ) and \
            issubclass( anyClass, DomainEventSubscriber ) and \
            anyClass != DomainEventSubscriber and  \
            cls.__getEventClass( anyClass ) == eventClass

File: preparation/src/shared/infrastructure/events/DomainEventSubscribersLocator.cs
/* 
 * 
 * Libraries
 *
*/

using System.Reflection;
using preparation.src.shared.domain;

/* 
 * 
 * Class
 *
*/

namespace preparation.src.shared.infrastructure {

    public class DomainEventSubscribersLocator { 

        /* 
         * 
         * Methods
         *
        */

        public static List<Type> locate( Type eventClass ) {
            // Variables
            List<Type>            subscriberClasses;
            Assembly              assembly;    
            IEnumerable<TypeInfo> satisfiedClasses;
            IEnumerable<TypeInfo> classTypes;
            IEnumerable<TypeInfo> implementedInterfaces;
            // Code
            assembly   = Assembly.Load( "preparation" );
            classTypes = assembly.GetExportedTypes().Select(
                i => i.GetTypeInfo()
            ).Where( i => i.IsClass && !i.IsAbstract );
            subscriberClasses = new List<Type>();
            foreach( TypeInfo classType in classTypes ) {
                implementedInterfaces = classType.ImplementedInterfaces.Select( i => i.GetTypeInfo() );
                satisfiedClasses = implementedInterfaces.Where( i =>
                    i.IsGenericType && 
                    i.GetGenericTypeDefinition() == typeof( DomainEventSubscriber<> ) &&
                    i.GenericTypeArguments.FirstOrDefault() == eventClass
                );
                if( satisfiedClasses.Count() != 0 ) {
                    subscriberClasses.Add( classType.AsType() );
                }
            }
            return subscriberClasses;
        }
    
    }

}